server/media/media.go

```diff
@@ -88,49 +89,59 @@ func matchCORSMethod(allowMethods []string, method string) bool {
 
 	method = strings.ToUpper(method)
 	for _, mm := range allowMethods {
-		if strings.ToUpper(mm) == method {
+		if mm == method {
 			return true
 		}
 	}
 
 	return false
 }
 
-// CORSHandler is the default preflight OPTIONS processor for use by media handlers.
+// CORSHandler is the default CORS processor for use by media handlers.
 func CORSHandler(req *http.Request, allowedOrigins []string, serve bool) (http.Header, int) {
-	if req.Method != http.MethodOptions {
-		// Not an OPTIONS request. No special handling for all other requests.
-		return nil, 0
-	}
-
-	var allowMethods []string
-	if serve {
-		allowMethods = []string{http.MethodGet, http.MethodHead, http.MethodOptions}
-	} else {
-		allowMethods = []string{http.MethodPost, http.MethodPut, http.MethodHead, http.MethodOptions}
-	}
-
 	headers := map[string][]string{
 		// Always add Vary because of possible intermediate caches.
-		"Vary":                             {"Origin", "Access-Control-Request-Method"},
-		"Access-Control-Allow-Headers":     {"*"},
-		"Access-Control-Max-Age":           {"86400"},
-		"Access-Control-Allow-Credentials": {"true"},
-		"Access-Control-Allow-Methods":     {strings.Join(allowMethods, ", ")},
+		"Vary": {"Origin", "Access-Control-Request-Method, Access-Control-Request-Headers"},
 	}
 
-	if !matchCORSMethod(allowMethods, req.Header.Get("Access-Control-Request-Method")) {
-		// CORS policy does not allow this method.
-		return headers, http.StatusNoContent
-	}
+	origin := req.Header.Get("Origin")
+
+	allowedOrigin := matchCORSOrigin(allowedOrigins, origin)
+	requestMethod := req.Header.Get("Access-Control-Request-Method")
+	if req.Method == http.MethodOptions && requestMethod != "" {
+		// Preflight request.
+
+		if allowedOrigin == "" {
+			return headers, http.StatusNoContent
+		}
+
+		var allowMethods []string
+		if serve {
+			allowMethods = []string{http.MethodGet, http.MethodHead, http.MethodOptions}
+		} else {
+			allowMethods = []string{http.MethodPost, http.MethodPut, http.MethodHead, http.MethodOptions}
+		}
+
+		if !matchCORSMethod(allowMethods, requestMethod) {
+			// CORS policy does not allow this method.
+			return headers, http.StatusNoContent
+		}
+
+		headers["Access-Control-Allow-Headers"] = []string{"*"}
+		headers["Access-Control-Allow-Credentials"] = []string{"true"}
```

Comment:
array?
The Access-Control-Allow-Credentials header expects a single string value, not a slice of strings